ethereumj-core/src/test/java/org/ethereum/sharding/processing/consensus/ShufflingCommitteeFactoryTest.java

```diff
@@ -0,0 +1,167 @@
+import static org.ethereum.sharding.processing.consensus.BeaconConstants.MIN_COMMITTEE_SIZE;
+import static org.ethereum.sharding.processing.consensus.BeaconConstants.SHARD_COUNT;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * @author Mikhail Kalinin
+ * @since 24.09.2018
+ */
+public class ShufflingCommitteeFactoryTest {
+
+    @Test
+    public void testCreate() {
+        int startShard = 3;
+        Committee[][] committees = new ShufflingCommitteeFactory().create(HashUtil.randomHash(),
+                IntStream.range(1, 3 * SHARD_COUNT * MIN_COMMITTEE_SIZE * CYCLE_LENGTH / 2).toArray(), startShard);
+
+        // each committee dedicated for own shard
+        int shard = startShard;
+        for (Committee[] slot : committees) {
+            for (Committee committee : slot) {
+                assertEquals(shard % SHARD_COUNT, committee.getShardId());
+                shard += 1;
+            }
+        }
+
+        // all committees are with the same shard
+        committees = new ShufflingCommitteeFactory().create(HashUtil.randomHash(),
+                IntStream.range(1, CYCLE_LENGTH + 1).toArray(), startShard);
+        for (Committee[] slot : committees) {
+            for (Committee committee : slot) {
+                assertEquals(startShard % SHARD_COUNT, committee.getShardId());
+            }
+        }
+    }
+
+    @Test
+    public void slotsPerShard() {
+        ShufflingCommitteeFactory factory = new ShufflingCommitteeFactory();
+        assertEquals(3, factory.calcShardsPerSlot(2 * MIN_COMMITTEE_SIZE * CYCLE_LENGTH * 2));
+        assertEquals(2, factory.calcShardsPerSlot(2 * MIN_COMMITTEE_SIZE * CYCLE_LENGTH));
+        assertEquals(1, factory.calcShardsPerSlot(CYCLE_LENGTH));
+        assertEquals(1, factory.calcShardsPerSlot(CYCLE_LENGTH - 1));
+        assertEquals(1, factory.calcShardsPerSlot(1));
+
+        assertEquals(CYCLE_LENGTH, factory.calcSlotsPerShard(1));
+        assertEquals(2, factory.calcSlotsPerShard(CYCLE_LENGTH * MIN_COMMITTEE_SIZE / 2));
+        assertEquals(4, factory.calcSlotsPerShard(CYCLE_LENGTH * MIN_COMMITTEE_SIZE / 3));
+        assertEquals(4, factory.calcSlotsPerShard(CYCLE_LENGTH * MIN_COMMITTEE_SIZE / 4));
+        assertEquals(8, factory.calcSlotsPerShard(CYCLE_LENGTH * MIN_COMMITTEE_SIZE / 7));
+        assertEquals(1, factory.calcSlotsPerShard(CYCLE_LENGTH * MIN_COMMITTEE_SIZE));
+        assertEquals(1, factory.calcSlotsPerShard(CYCLE_LENGTH * MIN_COMMITTEE_SIZE * 2));
+    }
+
+    @Test
+    public void testShuffle() {
+        int[] in = IntStream.range(1, BeaconConstants.MIN_COMMITTEE_SIZE * 1024).toArray();
+        int[] out = new ShufflingCommitteeFactory().shuffle(HashUtil.randomHash(), in);
+        assertFalse(Arrays.equals(in, out));
```

Comment:
It will fail one day 😿

Reply:
yep, in theory it may be, but hardly the case in practice :)